### ISSUE TITLE
contrib/util/custom-firewall.sh throws increasingly disconcerting messages

### ISSUE THREAD

**Issue by yebyen** [deis/deis#4281]
https://github.com/deis/deis/blob/master/contrib/util/custom-firewall.sh

I have a five-node cluster on DigitalOcean and I'm following according to the guide

http://docs.deis.io/en/latest/installing_deis/digitalocean/

Apply Security Group Settings

```
... [stuff]
# Accept ssh, http, https and git
-A Firewall-INPUT -m conntrack --ctstate NEW -m multiport -p tcp --dports 22,2222,80,443 -j ACCEPT

# Log and drop everything else
-A Firewall-INPUT -j LOG
-A Firewall-INPUT -j REJECT

COMMIT
Saving firewall Rules
Enabling iptables service
Loading custom iptables firewall
Created symlink from /etc/systemd/system/basic.target.wants/iptables-restore.service to /usr/lib64/systemd/system/iptables-restore.service.
iptables: No chain/target/match by that name.
Bad argument `ACCEPT'
Error occurred at line: 21
Done
Try `iptables-restore -h' or 'iptables-restore --help' for more information.
(stretch)yebyen@localhost:~/deis$ 
```

The command I'm using:
```
    for i in 1 2 3 4 5; do ssh core@deis-$i.[my.cluster.address] 'bash -s' < contrib/util/custom-firewall.sh; done
```

This output looks bad, if I had to guess I'd say my security group settings are probably not being (at least fully) applied.

**Comment by pwalsh** [comment]
I get the same: 3 node DO cluster.

**Comment by nathansamson** [comment]
Note that this symptom I have also seen with #4311  Root cause being that fleet is not able to contact the etcd cluster (and fleetctl list-machines isn't returning sensible output)

**Comment by carmstrong** [comment]
> Root cause being that fleet is not able to contact the etcd cluster

This is exactly what I saw with #4315. I had to manually `sudo systemctl start fleetd` on my hosts for things to come up. I haven't yet dug into figuring out what's going on, but it looks like there's a compatibility issue.

**Comment by nathansamson** [comment]
Note that in my tests fleet could not contact etcd because etcd wasn't running properly. Restarting fleetd wouldn't have helped much in that case I guess. You might have been hitting a slightly different problem?

**Comment by yebyen** [comment]
There are two problems in the original bug report, one of them is definitely related to etcd nodes not talking to each other.  The other one is somewhere around Bad argument 'ACCEPT' -- try it when etcd is talking then, I think you still get `No chain/target/match by that name.`

**Comment by aledbf** [comment]
@yebyen the script assumes that etcd is working in the cluster. If not the error that you see in the iptables rules is because here https://github.com/deis/deis/blob/master/contrib/util/custom-firewall.sh#L35 
$MACHINES_IP is empty. You can check that running `cat /var/lib/iptables/rules-save`.

**Comment by nathansamson** [comment]
Would it be worth to stop the script prematurely with a sensible warning if fleet isn't running properly? Would give users a better idea what to do

**Comment by bf4** [comment]
fwiw, I'm running into this too. It's driving me crazy. 1.9.1  I'm just going to paste https://raw.githubusercontent.com/carmstrong/deis/master/contrib/coreos/user-data.example and run `make discovery-url` again

btw, I'm running

```bash
deis_version=1.9.1
docl droplets | grep "deis-v${deis_version}-" | awk '{ print $8 }' | while read i ; do ssh "core@${i}" 'bash -s' < contrib/util/custom-firewall.sh ; done
```

**Comment by bf4** [comment]
oh, also, it's currently impossible to install the recommended coreos on digitalocean https://github.com/deis/deis/issues/4354#issuecomment-135822086

**Comment by yebyen** [comment]
Thanks for updating this ticket.  I have no other sources reporting a
problem with current CoreOS but I bumped into it myself, I think.

v1.9.1?

**Comment by yebyen** [comment]
I don't see v1.9.1 anymore in the git tags.  Was that branch abandoned?

I'm going back to try v1.9.0 again, that it seems to be gone from the tags archive is a good indicator that something was wrong with the point release.  FWIW I had successes with v1.9.0 after initially reporting failures, and I now think that I've been building it wrong ever since.  This issue subject still applies though, I think at least.

**Comment by yebyen** [comment]
```
iptables: No chain/target/match by that name.
```

Back to the subject, this was the line that had me worried, I've never seen it go away even when etcd is up and the other error-looking messages in the above snippet all don't show.  Does anyone know what line of the iptable script this message might refer to?  (Is there a misspelled -J target somewhere in that file?  If so, I think we can actually fix it and close this ticket.)

**Comment by yebyen** [comment]
By building it wrong, I mean I think I was building from master because there was no v1.9.1 to checkout in trunk, I'm using the client-go and checkout failed in $GOPATH so built from master instead of a tag

**Comment by yebyen** [comment]
I am not having any issue with v1.9.0 at all

**Comment by carmstrong** [comment]
> there was no v1.9.1 to checkout in trunk

Really? That's certainly odd - we've tagged it just like any other release.

**Comment by bf4** [comment]
@carmstrong My guess is that @yebyen needs to run `git fetch --tags` or something.  The tag is definitely there.. I've used it.. and here's the link https://github.com/deis/deis/tree/v1.9.1

**Comment by mboersma** [comment]
FWIW, I provisioned DigitalOcean from current master / v1.10.0 release candidate, and except for "No chain/target/match by that name" it looks to have applied correctly:
```console
$ for i in 1 2 3; do ssh core@deis-$i.deis.rocks 'bash -s' < contrib/util/custom-firewall.sh; done
Obtaining IP addresses of the nodes in the cluster...
Cluster IPs: 10.132.115.1,10.132.115.2,10.132.115.3
Creating firewall Rules...
*filter

:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
:Firewall-INPUT - [0:0]
-A INPUT -j Firewall-INPUT
-A FORWARD -j Firewall-INPUT
-A Firewall-INPUT -i lo -j ACCEPT
-A Firewall-INPUT -p icmp --icmp-type echo-reply -j ACCEPT
-A Firewall-INPUT -p icmp --icmp-type destination-unreachable -j ACCEPT
-A Firewall-INPUT -p icmp --icmp-type time-exceeded -j ACCEPT

# Ping
-A Firewall-INPUT -p icmp --icmp-type echo-request -j ACCEPT

# Accept any established connections
-A Firewall-INPUT -m conntrack --ctstate  ESTABLISHED,RELATED -j ACCEPT

# Enable the traffic between the nodes of the cluster
-A Firewall-INPUT -s 10.132.115.1,10.132.115.2,10.132.115.3 -j ACCEPT

# Allow connections from docker container
-A Firewall-INPUT -i docker0 -j ACCEPT

# Accept ssh, http, https and git
-A Firewall-INPUT -m conntrack --ctstate NEW -m multiport -p tcp --dports 22,2222,80,443 -j ACCEPT

# Log and drop everything else
-A Firewall-INPUT -j LOG
-A Firewall-INPUT -j REJECT

COMMIT
Saving firewall Rules
Enabling iptables service
Created symlink from /etc/systemd/system/basic.target.wants/iptables-restore.service to /usr/lib64/systemd/system/iptables-restore.service.
iptables: No chain/target/match by that name.
Loading custom iptables firewall
Done
...
```
We should investigate `iptables: No chain/target/match by that name.`

**Comment by mboersma** [comment]
> We should investigate `iptables: No chain/target/match by that name.`

This happens when running the rules for the first time; since they are meant to be idempotent, the script tries to flush custom rules that don't yet exist:
```sh
# Flush custom rules before the restore (so this script is idempotent)
sudo /usr/sbin/iptables -F Firewall-INPUT
```
So it's not an error and can be ignored. @aledbf may work on a PR to silence the error message in this case.

**Comment by yebyen** [comment]
Agreed that the scripts should be idempotent but I think that error still shows each time you re-run it.  I've been using "run the firewall script again" as a step to check that my DNS is correct and all cached copies of SSH host keys are current for all 5 nodes before moving on to the next step in the guide.

I'm not in front of the cluster, but I will try running the firewall script again later tonight to confirm this.

You're also right about `pull --tags` @b4f and @carmstrong -- that's what I was missing.  I had grabbed the tag from another machine and assumed it had been removed when I couldn't check it out as detached from last week's deisctl test machine, where I was using v1.9.0.  Do you get pull --tags when you pull master, or do you have to pull them both separately?  Sorry to add extra noise to this ticket.

**Comment by yebyen** [comment]
@mboersma is correct.  When the settings have been successfully applied, and the script is run again, no errors are printed.  You get errors from a node if it is not talking to the fleet correctly on etcd.

All such errors were able to be resolved in my most recent test on v1.10.0 by rebooting the malfunctioning node or running the script again.  Closing, seeing no further issue on my end.
